### DISCUSSION TITLE
Is it possible to reduce path finding time?

### DISCUSSION THREAD

**Discussion by NathanCQC** [NVIDIA/cuQuantum#64] [Q&A] [answered]
Hello,

I am contracting a circut with 25,000 tensors.

`path, info = network.contract_path(optimize={'samples': n_samples, 'slicing': {'min_slices': max(max_n_slices, size)}})
`
In this regime, the above function takes about 8000 seconds and the contraction is 1.57s.

Is it possible to make the path finding less expensive?

Changing n samples from 1-10 has no effect on this time. Would you be able to provide some insight there?

Thanks alot.

**Comment by haidarazzam** [comment]
Thank you for reaching out to the cuQuantum team, we are delighted to help resolve your issue. 
The pathfinding time is affected by the size of the network, the slicing, the number of samples and the number of threads. It is on average depending on the number of samples, for example using 1 thread 10 samples needs about 10 times the time of 1 sample, but 10 samples may requires almost same time as 1 sample if running on 10 or more threads.
What is the value of your min_slices? (Since the contraction time is too small, I would guess that no slicing is needed, usually we slice to make the contraction fit into the GPU but if the contraction time is small, it indirectly means that it is small contraction and fit into the GPU memory thus no need for slicing.
cuQuantum 23.06 (the version that is going to be released soon, has an automated option to reduce the pathfinder time based on the contraction time but at least there must be 1 sample (1 path) that need to be generated.
Options to reduce the time include:

1. disable slicing
2. reduce reconfIters (default is 500-1000 and can be reduced down to 200 100 to 0) 'CUTENSORNET_CONTRACTION_OPTIMIZER_CONFIG_RECONFIG_NUM_ITERATIONS'
3. reduce reconfLeafs to 6 'CUTENSORNET_CONTRACTION_OPTIMIZER_CONFIG_RECONFIG_NUM_LEAVES'

Can you share your network?

**Reply by NathanCQC** [reply]
Thanks for your detailed response:

You are right I should disable the slicing in this regime.

What does reconfIters do?

I am working on Perlmutter which has many GPU nodes. But I assume by thread here you mean each GPU node's gpu threads. It is not clear how to control this and I imagine must be handled somewhere in the C++ API.

Here are some benchmarks of my network. There are a few different benchmark sizes. It would be interesting to see the pathfinding and contraction times you manage to get and compare to my results. Here are the network in a pickled interleaved format.

You can open them with 

```
import pickle
n_sites = 2
with open(f'{n_sites}hubbard_hadamard_test.pkl', 'rb') as f:
    cuquantum_interleaved = pickle.load(f)
```

[hubbard_series_hadamard_test.zip](https://github.com/NVIDIA/cuQuantum/files/11982706/hubbard_series_hadamard_test.zip)

**Reply by yangcal** [reply]
Hello Nathan,

Thanks for uploading your cases. We're going to examine these networks and provide suggestions. 

Please see this [page](https://docs.nvidia.com/cuda/cuquantum/cutensornet/api/types.html?highlight=config_reconfig_num_iterations#_CPPv4N49cutensornetContractionOptimizerConfigAttributes_t64CUTENSORNET_CONTRACTION_OPTIMIZER_CONFIG_RECONFIG_NUM_ITERATIONSE) for documentation on reconflters

The threads here refer to CPU threads where path-finding is executed. 

'CUTENSORNET_CONTRACTION_OPTIMIZER_CONFIG_RECONFIG_NUM_ITERATIONS' is exposed under `cuquantum.cutensornet.configuration.ReconfigOptions.num_iterations`, you can simply set it up with

```python
optimize={'reconfiguration': {'num_iterations': 200}}
```

While we're examining these cases, it would also be helpful if you could upload the large cases, eg, the 8000s one that was problematic. We don't need the physical operand data, you can just pickle the shape and the indices for each tensor.

**Reply by NathanCQC** [reply]
Thanks, your help is greatly appreciated. 

What is the reconfiguration actually doing? I apologise for the questions.

The file 7hubbard_hadamard_test took 8000s, which is in that zip folder. I can send you even larger benchmarks if you would like?

**Reply by yangcal** [reply]
An introduction on our contraction path optimizer can be found [here](https://docs.nvidia.com/cuda/cuquantum/cutensornet/overview.html#contraction-optimizer). Reconfiguration can greatly improve the path qualify, especially in large tensor network but could also heavily increase the path-finding time. 

BTW, I tried to use the default path-finding option for your `n=7` case and path-finding only took 70s (snippet attached). Can you try it locally? 

```python
import pickle
import time

from cuquantum import contract, contract_path

def run_site(n_sites, *args, **kwargs):
    with open(f'{n_sites}hubbard_hadamard_test.pkl', 'rb') as f:
        cuquantum_interleaved = pickle.load(f)
    num_operands = len(cuquantum_interleaved[::2])
    t0 = time.perf_counter()
    path, info = contract_path(*cuquantum_interleaved)
    t1 = time.perf_counter() - t0
    print(f"{n_sites=}, {num_operands=}, {t1=}")
    print(info.slices)
    return

n_sites = 7
run_site(n_sites)
```

**Reply by NathanCQC** [reply]
LOL. You smoked me.

I don't have GPU locally. I will run your script on the Perlmutter loginnode and see what happens.

What set up are you running this on? Is it single GPU?

**Reply by leofang** [reply]
`contract_path()` currently does not use GPUs (thus using a CPU-only timer like `time.perf_counter()` is OK).

**Reply by yangcal** [reply]
I was running it on a DGXA100 station (single CPU and A100 card); as Leo mentioned, we currently only use CPU for path-finding. The memory limit (depending on which GPU you're going to use) may also affect the time it takes to find an acceptable path. In this case, I also tried `contract_path(*cuquantum_interleaved, options={'memory_limit': '32GB'})` (my testing A100 has 80GB) and it was still ~72s.

**Reply by NathanCQC** [reply]
This is still running... clearly, there is something wrong with the configuration. Running time >> 72s

**Reply by leofang** [reply]
Interesting... Just curious,
1. What's the environment in use? You can dump the output of `pip list` (or `conda list` if you use Conda on Perlmutter).
2. Any chance you have the MPI support turned on? (That is, you had `cutn.distributed_reset_configuration()` called prior to `contract_path()`.)

**Reply by NathanCQC** [reply]
I am running the small script just calculating the contract path above provided by @yangcal 

# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
boltons                   23.0.0                   pypi_0    pypi
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2023.5.7             hbcca054_0    conda-forge
certifi                   2023.5.7                 pypi_0    pypi
cffi                      1.15.1                   pypi_0    pypi
charset-normalizer        3.1.0                    pypi_0    pypi
cqlibs                    0.0.post1.dev122+gfc68fc8.d20230630          pypi_0    pypi
cryptography              41.0.1                   pypi_0    pypi
cuda-cudart               12.0.107             h59595ed_6    conda-forge
cuda-cudart_linux-64      12.0.107             h59595ed_6    conda-forge
cuda-nvrtc                12.0.76              h59595ed_1    conda-forge
cuda-nvtx                 12.0.76              hcb278e6_0    conda-forge
cuda-version              12.0                 hffde075_2    conda-forge
cupy                      12.1.0           py39hc7c1505_0    conda-forge
cuquantum-python          23.03.0          py39hcf40d7a_3    conda-forge
custatevec                1.3.0                h3a97aeb_3    conda-forge
cutensor                  1.7.0.1              hcb278e6_1    conda-forge
cutensornet               2.1.0           mpi_mpich_h9f53280_3    conda-forge
docopt                    0.6.2                    pypi_0    pypi
environs                  9.5.0                    pypi_0    pypi
fastrlock                 0.8              py39h5a03fae_3    conda-forge
h5py                      3.9.0                    pypi_0    pypi
idna                      3.4                      pypi_0    pypi
importlib-metadata        6.7.0                    pypi_0    pypi
inquanto                  2.0.0.post1.dev546+gfe032e6a2b9ed6          pypi_0    pypi
jaraco-classes            3.2.3                    pypi_0    pypi
jeepney                   0.8.0                    pypi_0    pypi
jinja2                    3.1.2                    pypi_0    pypi
keyring                   24.2.0                   pypi_0    pypi
lark-parser               0.12.0                   pypi_0    pypi
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libblas                   3.9.0           17_linux64_openblas    conda-forge
libcblas                  3.9.0           17_linux64_openblas    conda-forge
libcublas                 12.0.1.189           hcb278e6_2    conda-forge
libcufft                  11.0.0.21            hcb278e6_1    conda-forge
libcurand                 10.3.1.50            hcb278e6_0    conda-forge
libcusolver               11.4.2.57            hcb278e6_1    conda-forge
libcusparse               12.0.0.76            hcb278e6_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 13.1.0               he5830b7_0    conda-forge
libgfortran-ng            13.1.0               h69a702a_0    conda-forge
libgfortran5              13.1.0               h15d22d2_0    conda-forge
libgomp                   13.1.0               he5830b7_0    conda-forge
liblapack                 3.9.0           17_linux64_openblas    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libnvjitlink              12.0.76              hcb278e6_1    conda-forge
libopenblas               0.3.23          pthreads_h80387f5_0    conda-forge
libsqlite                 3.42.0               h2797004_0    conda-forge
libstdcxx-ng              13.1.0               hfd8a6a1_0    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libzlib                   1.2.13               hd590300_5    conda-forge
markupsafe                2.1.3                    pypi_0    pypi
marshmallow               3.19.0                   pypi_0    pypi
more-itertools            9.1.0                    pypi_0    pypi
mpi                       1.0                       mpich    conda-forge
mpi4py                    3.1.4                    pypi_0    pypi
mpich                     4.1.1                external_0    conda-forge
mpmath                    1.3.0                    pypi_0    pypi
msgpack                   1.0.5                    pypi_0    pypi
ncurses                   6.4                  hcb278e6_0    conda-forge
networkx                  3.1                      pypi_0    pypi
numpy                     1.23.5                   pypi_0    pypi
openssl                   3.1.1                hd590300_1    conda-forge
ordered-set               4.1.0                    pypi_0    pypi
packaging                 23.1                     pypi_0    pypi
pandas                    1.5.3                    pypi_0    pypi
pendulum                  2.1.2                    pypi_0    pypi
pip                       23.1.2             pyhd8ed1ab_0    conda-forge
pycparser                 2.21                     pypi_0    pypi
python                    3.9.16          h2782a2a_0_cpython    conda-forge
python-dateutil           2.8.2                    pypi_0    pypi
python-dotenv             1.0.0                    pypi_0    pypi
python-graphviz           0.20.1                   pypi_0    pypi
python_abi                3.9                      3_cp39    conda-forge
pytket                    1.16.0                   pypi_0    pypi
pytket-cutensornet        0.1.0                     dev_0    <develop>
pytz                      2023.3                   pypi_0    pypi
pytzdata                  2020.1                   pypi_0    pypi
qwasm                     1.0.1                    pypi_0    pypi
readline                  8.2                  h8228510_1    conda-forge
requests                  2.31.0                   pypi_0    pypi
scipy                     1.11.1                   pypi_0    pypi
secretstorage             3.3.3                    pypi_0    pypi
setuptools                68.0.0             pyhd8ed1ab_0    conda-forge
six                       1.16.0                   pypi_0    pypi
sourcedefender            10.0.13                  pypi_0    pypi
sympy                     1.12                     pypi_0    pypi
tgcrypto                  1.2.5                    pypi_0    pypi
tk                        8.6.12               h27826a3_0    conda-forge
types-pkg-resources       0.1.3                    pypi_0    pypi
typing-extensions         4.7.0                    pypi_0    pypi
tzdata                    2023c                h71feb2d_0    conda-forge
urllib3                   2.0.3                    pypi_0    pypi
wheel                     0.40.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
zipp                      3.15.0                   pypi_0    pypi

**Reply by NathanCQC** [reply]
Tomorrow I will investigate with my colleague who set up the env.

**Reply by NathanCQC** [reply]
I ran the run_site benchinging script using the container image on Perlmutter. Which they helped me write - see below.

```
shifter --image=docker:nathfitz/pytket-cutensornet-perlmutter-mpi python contract_path_time.py 
n_sites=7, num_operands=29587, t1=2808.997422230197
```

running from a container, the path finding is 40x slower than your benchmarks. Compared to x100 times slower with the raw conda installation.

Are you sure you ran 7 sites? 72s seems fast for a 30,000 tensor network. But maybe this is unsurprising to you.

Here is the shifter file.

```
FROM docker.io/nvidia/cuda:11.7.1-devel-ubuntu22.04

ENV DEBIAN_FRONTEND noninteractive

WORKDIR /opt

RUN \
    apt-get update && \
    apt-get upgrade --yes && \
        apt-get install --yes \
        build-essential \
        wget \
        git \
        gfortran \
        vim && \
    apt-get clean all

#install miniconda
#mpich requires python
ENV installer=Miniconda3-latest-Linux-x86_64.sh

RUN wget https://repo.anaconda.com/miniconda/$installer && \
/bin/bash $installer -b -p /opt/miniconda3 && \
rm -rf $installer

ENV PATH=/opt/miniconda3/bin:$PATH

#Install MPI
ARG mpich=4.1.1
ARG mpich_prefix=mpich-$mpich

RUN \
    wget https://www.mpich.org/static/downloads/$mpich/$mpich_prefix.tar.gz && \
    tar xvzf $mpich_prefix.tar.gz && \
    cd $mpich_prefix && \
    ./configure && \
    make -j 16 && \
    make install && \
    make clean && \
    cd .. && \
    rm -rf $mpich_prefix

RUN /sbin/ldconfig

#try getting rid of conda's linker, seems to help with building mpi4py
RUN rm -rf /opt/miniconda3/compiler_compat

RUN conda install cython -y

RUN python -m pip install mpi4py numpy cupy-cuda11x

ENV CUDA_PATH /usr/local/cuda/

RUN git clone https://github.com/NVIDIA/cuQuantum && \
    cd cuQuantum/python && \
    python -m pip install .
```

**Reply by yangcal** [reply]
Right I'm pretty sure that I'm running the largest problem (see attached). You're running it in the login node? I don't know much about Perlmutter, is it possible that the login node is restricted to non-intensive tasks? 
```n_sites=7, num_operands=29587, t1=71.19434936298057```

**Reply by NathanCQC** [reply]
On a compute node

using this submission script

#!/bin/bash

```
#SBATCH -N 1 
#SBATCH -C gpu
#SBATCH -G 1
#SBATCH -q regular
#SBATCH -t 01:00:00
#SBATCH -A m4147

#run the application:

export CUDA_VISIBLE_DEVICES=$SLURM_LOCALID

srun shifter --image=docker:nathfitz/pytket-cutensornet-perlmutter-mpi python3 contract_path_time.py

```

Speed is below

```
n_sites=7, num_operands=29587, t1=1743.6499659400433
((9586, 1), (14469, 1), (14213, 1), (-2728, 1), (2728, 1))
```


It is running on a single CPU thread I think. This is also the same for me when not using shifter.

I imagine this is quite hard for you to diagnose. I will bring it up with Perlmutter support aswell.

**Reply by leofang** [reply]
@NathanCQC I don't understand the dockerfile (?) that you posted earlier. It doesn't seem that cuQuantum got pulled when you build cuQuantum Python, so how was it done exactly? Specifically, this
```
RUN git clone https://github.com/NVIDIA/cuQuantum && \
    cd cuQuantum/python && \
    python -m pip install .
```
only builds cuQuanutm Python. But what DSOs was it linked to? How did it access cuQuantum symbols at build time?

**Reply by NathanCQC** [reply]
I did not write this unfortunately. The Perlmutter engineers did, I have notified them that we are discussing on this thread. They may be able to provide some feedback if they see this message. Do you expect that this will not install correctly?

As far as I am aware shifter substitutes some modules at run time, because it is not a true docker replacement. https://docs.nersc.gov/development/shifter/how-to-use/

**Reply by leofang** [reply]
> Do you expect that this will not install correctly?
> As far as I am aware shifter substitutes some modules at run time, because it is not a true docker replacement.

Yes. This could be concerning. It could be possible that an old (perhaps suboptimal for your use case) cuTensorNet got used. This is why I was asking for software versions etc earlier, which is now invalidated since you moved to use shifter.

Is it possible that you run on compute node, instead of the login node, with [the same env you had before](https://github.com/NVIDIA/cuQuantum/discussions/64#discussioncomment-6420849), just to satisfy my curiosity? 🙂

**Reply by NathanCQC** [reply]
Yes I have done both and the performance is the same. My colleague @yapolyak has also done this.

It is most likely something wrong with the slurm input.

**Reply by NathanCQC** [reply]
These are the packages in side the container when I start a bash script inside it

Package                 Version
----------------------- -----------
attrs                   23.1.0
boltons                 23.0.0
brotlipy                0.7.0
build                   0.10.0
CacheControl            0.12.14
certifi                 2023.5.7
cffi                    1.15.1
charset-normalizer      2.0.4
cleo                    2.0.1
conda                   23.5.0
conda-content-trust     0.1.3
conda-package-handling  2.0.2
conda_package_streaming 0.7.0
crashtest               0.4.1
cryptography            39.0.1
cupy-cuda11x            12.1.0
cuquantum-python-cu11   23.3.0
custatevec-cu11         1.3.0
cutensor-cu11           1.7.0
cutensornet-cu11        2.1.0
Cython                  0.29.33
distlib                 0.3.6
dulwich                 0.21.5
fastrlock               0.8.1
filelock                3.12.2
graphviz                0.20.1
html5lib                1.1
idna                    3.4
importlib-metadata      6.6.0
installer               0.7.0
jaraco.classes          3.2.3
jeepney                 0.8.0
Jinja2                  3.1.2
jsonpatch               1.32
jsonpointer             2.1
jsonschema              4.17.3
keyring                 23.13.1
lark-parser             0.12.0
lockfile                0.12.2
MarkupSafe              2.1.3
more-itertools          9.1.0
mpi4py                  3.1.4
mpmath                  1.3.0
msgpack                 1.0.5
networkx                3.1
numpy                   1.24.3
packaging               23.0
pandas                  2.0.2
pexpect                 4.8.0
pip                     23.0.1
pkginfo                 1.9.6
platformdirs            3.5.3
pluggy                  1.0.0
poetry                  1.5.1
poetry-core             1.6.1
poetry-plugin-export    1.4.0
ptyprocess              0.7.0
pycosat                 0.6.4
pycparser               2.21
pyOpenSSL               23.0.0
pyproject_hooks         1.0.0
pyrsistent              0.19.3
PySocks                 1.7.1
python-dateutil         2.8.2
pytket                  1.16.0
pytz                    2023.3
qwasm                   1.0.1
rapidfuzz               2.15.1
requests                2.28.1
requests-toolbelt       1.0.0
ruamel.yaml             0.17.21
ruamel.yaml.clib        0.2.6
scipy                   1.10.1
SecretStorage           3.3.3
setuptools              65.6.3
shellingham             1.5.0.post1
six                     1.16.0
sympy                   1.12
tomli                   2.0.1
tomlkit                 0.11.8
toolz                   0.12.0
tqdm                    4.65.0
trove-classifiers       2023.5.24
types-pkg-resources     0.1.3
typing_extensions       4.6.3
tzdata                  2023.3
urllib3                 1.26.15
virtualenv              20.23.0
webencodings            0.5.1
wheel                   0.38.4
zipp                    3.15.0
zstandard               0.19.0

**Reply by leofang** [reply]
Thanks for confirming, Nathan. I forgot that I designed `python -m pip install .` to pull our wheels from PyPI. So all is good regarding software env.

To your other reply earlier:
> It is running on a single CPU thread I think.

Did you use `htop` or other tools on the compute node to confirm this?

**Reply by NathanCQC** [reply]
I am looking into this now.

I am also running from inside this container https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuquantum-appliance

But I think the problem is still the same.

**Reply by NathanCQC** [reply]
top is never over 100% (Much lower infact most of the time)

**Comment by yapolyak** [comment]
Can I jump in @yangcal  and ask how do you control number of CPU threads you are running `contract_path` with? Are those logical threads of a single core or multiple cores? I presume hyperthreading might be off by default on Perlmutter. (I have Not run the example yet myself, let me try (now or tomorrow morning UK time)).

**Reply by yapolyak** [reply]
Ah, and how many threads have you used for your (fast!) benchmark calculation?

**Reply by yapolyak** [reply]
And what CPU do you use :D

**Reply by yapolyak** [reply]
I found this https://docs.nvidia.com/cuda/cuquantum/python/api/generated/cuquantum.cutensornet.ContractionOptimizerConfigAttribute.html?highlight=thread#cuquantum.cutensornet.ContractionOptimizerConfigAttribute.HYPER_NUM_THREADS looks like by default it should use 31 hyperthreads?

**Reply by yapolyak** [reply]
Ah no, "Default is number-of-cores / 2" but what defines number of cores? All available?

**Reply by yapolyak** [reply]
Ah, and here it is exposed, I see https://docs.nvidia.com/cuda/cuquantum/python/api/generated/cuquantum.OptimizerOptions.html?highlight=thread

**Reply by yapolyak** [reply]
I am sorry I now see you mentioned you only used single CPU... So, a single logical thread as well? This is really strange - I can confirm it takes ages on Perlmutter. I am now trying to use multiple cores.
I hope this discrepancy is not due to how we install cuQuantum (or perhaps I should hope otherwise indeed). We use the following commands (specific to Perlmutter in the sense that we need to use system's MPICH library):
```
conda create -n py-cuquantum-23.03.0-mypich-py3.9 -c conda-forge python=3.9 cuquantum-python "mpich=*=external_*"
# Update pip
pip install --upgrade pip
# Setup MPI
MPICC="cc -shared" pip install --force-reinstall --no-cache-dir --no-binary=mpi4py mpi4py
```

**Reply by yapolyak** [reply]
I do recall I had some cuQuantum tests crashing a while ago but I got lazy afterwards and have not checked them ever since, as in general we could run calculation OK.

**Reply by leofang** [reply]
I believe @yangcal meant to say he used [the provided script](https://github.com/NVIDIA/cuQuantum/discussions/64#discussioncomment-6410199) as-is, so it's not just single-thread; `HYPER_NUM_THREADS` (= `OptimizerOptions.threads` as you noted) is default to the number of hyperthreads / 2 if this option is not set. So the machine shouldn't be fully saturated by default.

**Reply by leofang** [reply]
@yapolyak I think [the way your installed it](https://github.com/NVIDIA/cuQuantum/discussions/64#discussioncomment-6430716) should be fine. But note that Yang's script does not need MPI, so we should rule it out: What happens if you run in an environment without MPI? Do you still see slowness?

**Reply by yapolyak** [reply]
That's an interesting point @leofang - I will try this tomorrow! I am trying to use more than a single core, but according to `top`, cuQuantum is only using one... Perhaps I am doing something wrong with Slurm, although I don't think so...

**Reply by yapolyak** [reply]
Just to make sure - one should not use environment variables such as $OMP_NUM_THREADS?

**Reply by yapolyak** [reply]
Right I am telling Slurm to use 64 cores/threads (`-c 64`), explicitly telling cuQuantum to use 64 threads:
```
path, info = contract_path(*cuquantum_interleaved, optimize=OptimizerOptions(threads=64))
```
But it is still using only 100% CPU... I guess we need to ask Perlmutter people why that could be...

**Reply by yapolyak** [reply]
Just in case, I've set $OMP_NUM_THREADS to 64, still nothing. Tomorrow I will try a non-MPI installation. Many thanks and have a nice day!

**Reply by leofang** [reply]
As Yang alluded to in his other post, `OMP_NUM_THREADS` has no effect as we don't use OpenMP as the threading layer in cuTensorNet.

**Comment by yangcal** [comment, marked as answer]
Hello @NathanCQC @yapolyak ,

We've pinned down a couple of factors and hopefully that can resolve all the questions above:

1.	Regarding the large performance difference in path-finding, turns out the `cuquantum-python` that we're testing with is our soon to-be-released 23.06 version (apologies here). We've added a few optimizations in our path-finder and that should be where the big difference is using my snippet. I tried the snippet again with 2303 and it was also stuck in the path-finding stage for at least tens of minutes. 
2.	For large tensor network like you have with n=7, you're encouraged to turn on hyper sampling via `OptimizerOptions.samples`(default is off). You can also adjust the number of threads via `OptimizerOptions.threads` or use the default half of cores. There is no need to specify OMP_NUM_THREADS for threading. Note that, threading kicks in after the first sample is finished. For your large network, the first sample itself is taking long time to complete and that's why you haven't observed the increased CPU utility. In 2306, there is also a new feature that further improves our path-finder when hyper sampling is on so stay tuned.

Hope that resolves all the questions above. We apologize for some negligence on our side and hope to exchange more feedbacks once our 2306 release is out.

**Reply by NathanCQC** [reply]
Thanks for figuring this out for us! Good to know it is not our problem. It's pretty exciting that you managed to get such a performance increase. Do you think there is scope for even more improvement in the future? This is very promising for our future development with your library.

I notice the version number on pip is 23.06. Would we be able to use new optimisiations if we install from pip.

If not when do you expect the new release to be out? Also can you say how you managed to achieve such a big improvement in performance?

BTW we are UK based hence why we could not respond to your message yesterday.

Can we get a similar performance by changing the OptimiserOptions inputs?

Kind Regards
Nathan

**Reply by yapolyak** [reply]
Ah, this is great news (and big relief), thank you very much for figuring this out, and no problem! :)

The fact that multiple threads spin up only after the first sample is done is probably explaining why I was only seeing 100% CPU load for the first several minutes of the run! 

Out of curiosity, what do you use for thread parallelism (in case you can share this info of course)?

We are looking forward to use 23.06 indeed!

**Reply by yangcal** [reply]
We're always looking for opportunities to optimize all functionalities in cuquantum and you shall be able to upgrade to the new version with all new features using pip/conda in a few days. We've improved our internal path searching algorithm so you should be able to observe the speedup automatically, eg, even with default setting like in my snippet. Threading parallelism was used in our hyper optimizer, see [here](https://docs.nvidia.com/cuda/cuquantum/cutensornet/overview.html#hyper-optimizer) for detailed explanation.

**Reply by leofang** [reply]
btw 23.06 has been fully posted. Due to the complexity of our offering and release process, it may not be always possible to get all things done at once, and you did observe this (pip wheels were out earlier than other pieces). @NathanCQC @yapolyak Feel free to give it a shot and let us know how it goes! 🙂

**Reply by NathanCQC** [reply]
Wow, you guys turned this around very quickly. Thanks a lot. I will let you know and run the benchmarking script today.

I will also try the auto parallelism.

**Reply by yapolyak** [reply]
Great news! We will give it a go! Thank you very much for your work and support!

**Reply by NathanCQC** [reply]
I see x20 speed up. For the 7 site hubbard model.

**Reply by yangcal** [reply]
The x20 speedup was compared with the 1743s that you previously observed on a compute node? That amounts to ~80 seconds and is about on par with our internal benchmark.
